### ISSUE TITLE
require is not defined with es6 native modules => wrong import path and require fonction in webpack output

### ISSUE THREAD

**Issue by rvion** [pmmmwh/react-refresh-webpack-plugin#294]
I get a 

🔴 `ReferenceError` `require is not defined`

📝I use esm native modules.
📝My config matches the readme

but webpack generate a weird import line for `__webpack_require__.$Refresh$.runtime` for every module:

```js
                            WRONG REQUIRE v                         WRONG PATH v
__webpack_require__.$Refresh$.runtime = require('C:/Users/myusername/dev/devops-cli/node_modules/react-refresh/runtime.js');
```

🔍it is as if it was not using the correct require, nor rewriting properly the path.

You can find more details below as well as a longer extract from webpack output.
What else can I provide ?

Any help or workaround appreciated ! 👍 

-------------------

config: 
```
plugins: [
    new webpack.HotModuleReplacementPlugin(),
    new ReactRefreshWebpackPlugin(),
]
....
module: {
    rules: [
        {
            // https://github.com/pmmmwh/react-refresh-webpack-plugin
            test: /\.js$/,
            exclude: /node_modules/,
            use: [
                {
                    loader: require.resolve('babel-loader'),
                    options: { plugins: [require.resolve('react-refresh/babel')], },
                },
            ],
        },
    ]
}
```

misc packages versions:

- "@pmmmwh/react-refresh-webpack-plugin": "0.4.3",
...
- "@babel/core": "7.12.10",
- "@babel/plugin-proposal-decorators": "^7.12.12",
- "@babel/plugin-transform-react-jsx": "^7.12.12",
- "@babel/preset-react": "^7.12.10",
- "@babel/preset-typescript": "^7.12.7",
...

        
```js
/***/ "./packages/test/counterUI1.js":
/*!*************************************!*\
  !*** ./packages/test/counterUI1.js ***!
  \*************************************/
/***/ ((__unused_webpack___webpack_module__, __webpack_exports__, __webpack_require__) => {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export */ __webpack_require__.d(__webpack_exports__, {
/* harmony export */   "CounterUI": () => /* binding */ CounterUI,
/* harmony export */   "foo": () => /* binding */ foo
/* harmony export */ });
/* harmony import */ var react__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! react */ "./node_modules/react/index.js");
/* harmony import */ var _emotion_react__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(/*! @emotion/react */ "./node_modules/@emotion/react/dist/emotion-react.browser.esm.js");
/* provided dependency */ var __react_refresh_utils__ = __webpack_require__(/*! ./node_modules/@pmmmwh/react-refresh-webpack-plugin/lib/runtime/RefreshUtils.js */ "./node_modules/@pmmmwh/react-refresh-webpack-plugin/lib/runtime/RefreshUtils.js");
/* provided dependency */ var __react_refresh_error_overlay__ = __webpack_require__(/*! ./node_modules/@pmmmwh/react-refresh-webpack-plugin/overlay/index.js */ "./node_modules/@pmmmwh/react-refresh-webpack-plugin/overlay/index.js");
                                            v HERE v 
__webpack_require__.$Refresh$.runtime = require('C:/Users/rvion/dev/devops-cli/node_modules/react-refresh/runtime.js');
__webpack_require__.$Refresh$.setup(module.id);
```

Thanks

**Comment by pmmmwh** [comment]
Hi, thanks for reporting this bug. I think this has something to do with our way of injecting dependencies via a loader. I can publish a beta version that might fix this, would you mind trying it out to see if it works or provide a minimal native ESM webpack project for reproduction?

**Comment by SamMorrowDrums** [comment]
@pmmmwh I would happily. I'm having same issue.

**Comment by rvion** [comment]
thanks @pmmmwh ! I can't test easily right now, since I reverted most of my changes related to migrating to esnext. (Too many things broken across various utilities I use) Thanks anyway for the quick feedback.

thanks @SamMorrowDrums for stepping in !

**Comment by 16patsle** [comment]
I'm also seeing this issue, with a relatively minimal webpack config. Everything should be up to date, since I set up this project only a few days ago. Let me know if there's anything I can do to help.

<details>
<summary>webpack.config.js</summary>

```javascript
const path = require("path");
const HtmlWebPackPlugin = require('html-webpack-plugin')
const ReactRefreshWebpackPlugin = require('@pmmmwh/react-refresh-webpack-plugin')

const createConfig = isDevelopment => ({
  entry: path.resolve(__dirname, 'src/index.js'),
  //target: isDevelopment ? 'web' : 'browserslist',
  output: {
    path: path.resolve(__dirname, "dist"),
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
          options: {
            plugins: [
              isDevelopment && require.resolve('react-refresh/babel'),
            ].filter(Boolean),
          },
        },
      },
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader', 'postcss-loader'],
      },
    ],
  },
  plugins: [
    new HtmlWebPackPlugin({
      template: './src/index.html',
    }),
    isDevelopment && new ReactRefreshWebpackPlugin(),
  ].filter(Boolean),
})

module.exports = (env, argv) => createConfig(argv.mode === 'development')

```

<details>
<summary>Also .babelrc, just in case</summary>

```json
{
  "presets": [
    [
      "@babel/preset-env",
      {
        "bugfixes": true,
        "useBuiltIns": "usage",
        "corejs": "3.8",
        "shippedProposals": true
      }
    ],
    "@babel/preset-react"
  ]
}

```
</details>

</details>

**Comment by pmmmwh** [comment]
Hi, I'm implementing a fix for this issue. Just so I can understand the issue more, are you all running native ESM with Webpack 5+? My impression is that the fix might not work quite well in Webpack 4 if you run strict ESM (where `module` is not available).

**Comment by 16patsle** [comment]
Thanks. I'm using the latest version of Webpack 5 yes.

**Comment by jeremytenjo** [comment]
Any updates with this issue?

**Comment by pmmmwh** [comment]
A fix should be ready this week. I'm planning to ship 0.5.0 within February.

**Comment by jeremytenjo** [comment]
> A fix should be ready this week. I'm planning to ship 0.5.0 within February.

Great! Thank you 👍

**Comment by xiaoshuangLi** [comment]
I got same error, because i defined ```require``` function.

<img width="615" alt="WechatIMG3416" src="https://user-images.githubusercontent.com/12060054/110427058-18162200-80e2-11eb-8e6b-bba9cde7b968.png">

**Comment by pmmmwh** [comment]
Hi, unfortunately the semantics needed for the fix is a bit more complex than expected.

The runtime is currently in a half-baked state where some tests pass and some don't (hopefully some progress in a PR soon). I'm continuously working on it currently.

**Comment by sokra** [comment]
> * currently fixed but unreleased

I'll try to release this today. The error semantics are the following with this change:

* The dispose handler of the old module is called.
* The new module factory is called (module evaluation)
* Any error during module evaluation will lead to an "module" in error state in the module cache. `module.error` is set to this error.
  * This also applies for compilation errors when `optimization.emitOnErrors: true` is set.
  * Note that the module might have partially executed until the point of the error!!
  * This is a new module, which is !== the old module.
* A self accept error handler of the old module is called with the runtime error that occured.
  * It also gets access to the new module via the `module` property in the second argument
  * It might use the HMR api of the new module `module.hot`
* The new module might or might not have self accepted.
  * It depends if `accept` was called before the error occured
  * Even if the `accept` call is the first statement, it might not call it, when compilation errors or parse errors occur.
  * There can only be one self accept error handler. Calling `accept` twice will override it.
* The new module might or might not have registered a `dispose` handler.
  * Some like above
  * There can be multiple dispose handlers. Calling `addDisposeHandler` twice will add multiple.
  * `removeDisposeHandler` can be used to remove handlers
  * You can use `import.meta.webpackHot.data` to store state in module evaluation and access this later via `module.hot.data` in the self accept error handler.
* If you self accept the new module in the self accept error handler, make sure to also register a dispose handler to not loose `data`.
* It's now possible to handle errors in `import.meta.webpackHot.accept("./dep", ...)` too, but I think you don't use that anyway.

**Comment by swissspidy** [comment]
I there any update on this, anything people can help with?

**Comment by pmmmwh** [comment]
I'll try to push out the WIP new runtime on this tonight.

**Comment by ntucker** [comment]
I'm experiencing this problem in the latest rc with a package that uses the new package.json 'exports' member to define es6 modules.

https://github.com/coinbase/rest-hooks/pull/996
yarn install
cd examples/todo-app
yarn start

**Comment by 2heal1** [comment]
> I got same error, because i defined `require` function.
> 
> <img alt="WechatIMG3416" width="615" src="https://user-images.githubusercontent.com/12060054/110427058-18162200-80e2-11eb-8e6b-bba9cde7b968.png">
so do i,have you solved this problem?

**Comment by pmmmwh** [comment]
> so do i,have you solved this problem?

Rename the function. `require` is a global in the CommonJS module system, there's nothing we can do about it.

**Comment by ipsips** [comment]
i can confirm that this is still broken in `0.5.0-rc.5 `

<img width="825" alt="Screenshot 2021-08-28 at 12 20 59" src="https://user-images.githubusercontent.com/1770542/131213120-f541e812-6ba1-4059-bc61-ea30f67a1df2.png">

**Comment by pmmmwh** [comment]
> i can confirm that this is still broken in `0.5.0-rc.5 `

This has something to do with monorepo compatibility, it will be addressed in the next RC.

**Comment by ntucker** [comment]
Thanks @pmmmwh that's good to hear! <3

**Comment by almir-campos** [comment]
I'm having the same problem. 

After some debuging I found that the problem first occurs in the file `node_modules/webpack/hot/emitter.js` at its very first line:

```
var EventEmitter = require("events");   // <--The problem first happens here
module.exports = new EventEmitter();
```
Obviously, I hacked this file this way:
```
// var EventEmitter = require("events"); 
import EventEmitter from "events"
export default new EventEmitter();
```
Then, fortunately, the error _'required is not defined_' was gone in this file but, unfortunately, it occurred in the file `./dev-server.js`, in the same directory.

It was clearly begging me for another hacking, what I did in this way:

```
// var hotEmitter = require("./emitter");
import('./emitter')
	  .then( hotEmitter => {
	     hotEmitter.on("webpackHotUpdate", function(currentHash) {
	    	lastHash = currentHash;
		if (!upToDate() && module.hot.status() === "idle") {
			log("info", "[HMR] Checking for updates on the server...");
			check();
		}
             });
	     log("info", "[HMR] Waiting for update signal from WDS...");
	  });
```
Again, the error was gone in that file but, to my despair, the error moved to another place related to a `require chunk loading` thing.

 I guess that this has to be with a comment made by @pmmmwh above:

> unfortunately the semantics needed for the fix is a bit more complex than expected.

At the end, I'm under the impression that the code is a little "spaghettified" in this module - which represents a good opportunity for improvements - and that wouldn't be very easy to fix the problem.

Does it make sense? Please, let me know.

**Comment by pmmmwh** [comment]
> I'm having the same problem. 
> 
> 
> 
> After some debuging I found that the problem first occurs in the file `node_modules/webpack/hot/emitter.js` at its very first line:

This sounds like either a bug on Webpack's side (what version are you using?) or a misconfiguration with your ESM setup. These files are internals of Webpack HMR and should not have to be touched for things to work.

**Comment by Param1raj** [comment]
Did you find the required configurations of webpack or ESM setup to make this work?

**Comment by boycce** [comment]
> > i can confirm that this is still broken in `0.5.0-rc.5 `
> 
> This has something to do with monorepo compatibility, it will be addressed in the next RC.

I just started using npm workspaces and encountered this issue, did this get resolved?

```
Uncaught ReferenceError: require is not defined
    at ../util.js (index.js:42:1) 
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:22:1)
    at fn (hot module replacement:61:1)
    at ../client/app.js (floating-ui.utils.mjs:138:362)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:22:1)
    at fn (hot module replacement:61:1)
    at ../client/index.js (app.js:300:2)
    at options.factory (react refresh:6:1)
```

runtime/react refresh
```
__webpack_require__.i.push((options) => {
	const originalFactory = options.factory;
	options.factory = function (moduleObject, moduleExports, webpackRequire) {
		__webpack_require__.$Refresh$.setup(options.id);
		try {
			originalFactory.call(this, moduleObject, moduleExports, webpackRequire);    <------- line 6
		} finally {
			if (typeof Promise !== 'undefined' && moduleObject.exports instanceof Promise) {
				options.module.exports = options.module.exports.then(
					(result) => {
						__webpack_require__.$Refresh$.cleanup(options.id);
						return result;
					},
					(reason) => {
						__webpack_require__.$Refresh$.cleanup(options.id);
						return Promise.reject(reason);
					}
				);
			} else {
				__webpack_require__.$Refresh$.cleanup(options.id)
			}
		}
	};
})

**Comment by pmmmwh** [comment]
This should be fixed in v0.6.0.
